includes/class-fs-garbage-collector.php

```diff
@@ -0,0 +1,458 @@
+        }
+    }
+
+    class FS_Plugin_Garbage_Collector extends FS_A_Product_Garbage_Collector {
+        function __construct( FS_Options $_accounts ) {
+            $this->_accounts = $_accounts;
+        }
+
+        function get_option_names()
+        {
+            return array(
+                'sites',
+                'plans',
+                'plugins',
+                'active_plugins',
+                'all_plugins',
+            );
+        }
+
+        function get_type() {
+            return WP_FS__MODULE_TYPE_PLUGIN;
+        }
+
+        function get_plural_type() {
+            return WP_FS__MODULE_TYPE_PLUGIN . 's';
+        }
+    }
+
+    class FS_Theme_Garbage_Collector extends FS_A_Product_Garbage_Collector {
+        function __construct( FS_Options $_accounts ) {
+            $this->_accounts = $_accounts;
+        }
+
+        function get_option_names()
+        {
+            return array(
+                'theme_sites',
+                'theme_plans',
+                'themes',
+                'active_themes',
+                'all_themes',
+            );
+        }
+
+        function get_type() {
+            return WP_FS__MODULE_TYPE_THEME;
+        }
+
+        function get_plural_type() {
+            return WP_FS__MODULE_TYPE_THEME . 's';
+        }
+    }
+
+    class FS_Product_Garbage_Collector implements FS_I_Garbage_Collector {
+
+        /**
+         * @var FS_A_Product_Garbage_Collector[]
+         */
+        private $_products_garbage_collector = array();
+
+        public function __construct( FS_Options $_accounts ) {
+            $this->add_garbage_collector( new FS_Plugin_Garbage_Collector( $_accounts ) );
+            $this->add_garbage_collector( new FS_Theme_Garbage_Collector( $_accounts ) );
+        }
+
+        private function add_garbage_collector( FS_A_Product_Garbage_Collector $garbage_collector ) {
+            $this->_products_garbage_collector[ $garbage_collector->get_type() ] = $garbage_collector;
+        }
+
+        function clean() {
+            $has_cleaned = false;
+
+            foreach ( $this->_products_garbage_collector as $product_garbage_collector ) {
+                if ( $product_garbage_collector->clean() ) {
+                    $has_cleaned = true;
+                }
+            }
+
+            return $has_cleaned;
+        }
+
+        function get_types() {
+            return array_keys( $this->_products_garbage_collector );
+        }
+    }
+
+    class FS_User_Garbage_Collector implements FS_I_Garbage_Collector {
+        private $_accounts;
+
+        private $_types;
+
+        public function __construct( FS_Options $_accounts, array $types ) {
+            $this->_accounts = $_accounts;
+            $this->_types    = $types;
+        }
+
+        public function clean() {
+            $users = Freemius::get_all_users();
+
+            $user_has_install = array();
+
+            foreach ( $this->_types as $product_type ) {

```

Comment:
@fajardoleo Proposing to extract this to a private function for readability and rename it to `$user_has_install_map` or something like that:

```
$user_has_install_map = $this->get_user_has_install_map();
```